x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/TimeSeriesBlockHash.java

```diff
@@ -0,0 +1,155 @@
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.util.BitArray;
+import org.elasticsearch.common.util.BytesRefHash;
+import org.elasticsearch.common.util.LongLongHash;
+import org.elasticsearch.compute.aggregation.AggregatorMode;
+import org.elasticsearch.compute.aggregation.GroupingAggregatorFunction;
+import org.elasticsearch.compute.aggregation.SeenGroupIds;
+import org.elasticsearch.compute.data.Block;
+import org.elasticsearch.compute.data.BytesRefBlock;
+import org.elasticsearch.compute.data.BytesRefVector;
+import org.elasticsearch.compute.data.IntVector;
+import org.elasticsearch.compute.data.LongBlock;
+import org.elasticsearch.compute.data.LongVector;
+import org.elasticsearch.compute.data.Page;
+import org.elasticsearch.compute.operator.DriverContext;
+import org.elasticsearch.core.Releasables;
+
+import java.util.Objects;
+
+public final class TimeSeriesBlockHash extends BlockHash {
+
+    private final AggregatorMode mode;
+    private final int tsHashChannel;
+    private final int timestampChannel;
+    private final Rounding.Prepared preparedRounding;
+    private final BytesRefHash tsidHashes;
+    private final LongLongHash intervalHash;
+
+    long groupOrdinal = -1;
+    BytesRef previousTsidHash;
+    long previousTimestamp;
+
+    public TimeSeriesBlockHash(
+        AggregatorMode mode,
+        int tsHashChannel,
+        int timestampChannel,
+        Rounding rounding,
+        DriverContext driverContext
+    ) {
+        super(driverContext.blockFactory());
+        this.mode = mode;
+        this.tsHashChannel = tsHashChannel;
+        this.timestampChannel = timestampChannel;
+        this.preparedRounding = rounding != null ? rounding.prepareForUnknown() : null;
+        this.tsidHashes = new BytesRefHash(1, blockFactory.bigArrays());
+        this.intervalHash = new LongLongHash(1, blockFactory.bigArrays());
+    }
+
+    @Override
+    public void close() {
+        Releasables.close(tsidHashes, intervalHash);
+    }
+
+    @Override
+    public void add(Page page, GroupingAggregatorFunction.AddInput addInput) {
+        BytesRefBlock tsHashBlock = page.getBlock(tsHashChannel);
+        BytesRefVector tsHashVector = Objects.requireNonNull(tsHashBlock.asVector());
+        try (var ordsBuilder = blockFactory.newIntVectorBuilder(tsHashVector.getPositionCount())) {
+            LongBlock timestampIntervalBlock = page.getBlock(timestampChannel);
+            BytesRef spare = new BytesRef();
+            if (preparedRounding != null) {
+                for (int i = 0; i < tsHashVector.getPositionCount(); i++) {
+                    BytesRef tsHash = tsHashVector.getBytesRef(i, spare);
+                    long timestampInterval = preparedRounding.round(timestampIntervalBlock.getLong(i));
+                    if (tsHash.equals(previousTsidHash) == false || timestampInterval != previousTimestamp) {
+                        long tsidOrdinal = tsidHashes.add(tsHash);
```

Comment:
I think we can avoid using these two hashes since data arrives in order in partial mode. Let's keep it simple for now.

Reply:
Good point, in partial mode we can group like this PR did yesterday (keep track of last seen tsid hash or tsid hash and time stamp and then increment a counter). I can add basically another implementation in a follow-up pr for this later.